Add Script to Remove Deleted REDCap Records from Cache

bencap opened the pull request:
When REDCap records are deleted, their associated entries must also be deleted
from the musher cache to ensure participants get directed to the correct
surveys. This was previously done manually, but if we parse the REDCap logs
to look for deleted records it is possible to do this as a nightly cleanup job
and not have to bother with manual removal.

Since REDCap logs don't have the netid (PHI!), we only have the record id of
the deleted record and have to find the netid to delete the cache record
somewhat inefficiently. I think this should be fine given when this job runs and
the current amount of cached netids. Deleted record ids are logged out to the
default musher log location at the info level.

bencap committed:
Add Script to Remove Deleted REDCap Records from Cache

When REDCap records are deleted, their associated entries must also be deleted
from the musher cache to ensure participants get directed to the correct
surveys. This was previously done manually, but if we parse the REDCap logs
to look for deleted records it is possible to do this as a nightly cleanup job
and not have to bother with manual removal.

Since REDCap logs don't have the netid (PHI!), we only have the record id of
the deleted record and have to find the netid to delete the cache record
somewhat inneficiently. I think this should be fine given when this job runs and
the current amount of cached netids. Deleted record ids are logged out to the
default musher log location at the info level.

bencap commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Musher/Lead Dawgs cache: figure out a process to maintain it to deal with deleted REDCap records](https://trello.com/c/M9ndgIdO/1002-musher-lead-dawgs-cache-figure-out-a-process-to-maintain-it-to-deal-with-deleted-redcap-records)

kschwabefry commented on bin/clean-cache:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    response = requests.post(LazyObjects.get_project().api_url, data=data)
+    response.raise_for_status()
```
You may want to additionally check for 200 status with an error message wrapped as html. (Its not the correct way for the api to respond, but its an issue weve noticed with the redcap api). Check out  the redcap module in id3c for an example of how it handles this.
Additionally - what are your thoughts on incorporating this into that core redcap module?

bencap replied:
Ah yup, makes sense and I can add that check. I could incorporate it into the core module, but it might make more sense broken out as a function in the redcap utils module within Musher depending on how much we think we want to programmatically access REDCap logs. 

Within the core module though it would be cool to have some functions that get logging like we already have for records. I'd want it to be a more robust function than here though with some additional arguments and checks.

bencap commented on bin/clean-cache:
```diff
@@ -0,0 +1,57 @@
+    deleted_ids = [int(deleted_record['details'].split('=')[1].replace("'", "").strip()) for deleted_record in records]
+    cached_records = map(LazyObjects.get_cache().get, LazyObjects.get_cache())
+
+    # unfortunately we use the netid to cache records not the record_id, so we have to
+    # find deleted record ids naively within the cache to get the associated netid.
+    for record in cached_records:
+        if int(record['record_id']) in deleted_ids:
+            LazyObjects.get_cache().delete(record['netid'])
+            LOG.info(f'Deleted record with id<{record["record_id"]}> from the musher cache.')
```
Yeah I'll simplify this comment. Mostly wanted to make it explicit that we are caching by netid for the next person who comes along.

bencap committed:
Add Functions to Get REDCap Logs and Validate Responses to REDCap Utils

This change adds two functions to the REDCap utilities package. The first is gets
logging data for deleted records based on a start and end time. This will be used
to do a nightly cleanup of the Musher cache. The second is a wrapper for REDCap
post requests, which can be notoriously finicky. Sometimes the API will return
a 200 code but not the correct response value. We don't want our cache cleanup to
fail silently, so we check to make sure the response we expect is within the
actual response before using it.

kschwabefry reviewed (approved):
Thanks for adding the additional logging and safety checks!

bencap commented:
Deployed.
